network/network.go

```diff
@@ -874,23 +881,47 @@ func (n *network) gossipContainer(chainID, containerID ids.ID, container []byte)
 
 // assumes the stateLock is held.
 func (n *network) track(ip utils.IPDesc) {
-	if n.closed.GetValue() {
+	str := ip.String()
+	if !n.shouldTrack(str) {
 		return
 	}
 
+	// Add [str] to disconnectedIPs to signal to the new goroutine
+	// not to terminate.
+	n.disconnectedIPs[str] = struct{}{}
+	go n.connectTo(ip)
+}
+
+// assumes the stateLock is held
+// return true if [ip] should be tracked
+func (n *network) shouldTrack(str string) bool {
+	// If [str] is already being tracked, do not attempt to track it
+	_, isDisconnected := n.disconnectedIPs[str]
+	// If the network is already connected to [str], do not attempt to track it
+	_, isConnected := n.connectedIPs[str]
+	// If [str] is registered as one of my IPs, do not attempt to track it
+	_, isMyself := n.myIPs[str]
+	// If the network is closed, do not attempt to track any new IPs
+	closed := n.closed.GetValue()
+
+	return !closed && !isDisconnected && !isConnected && !isMyself
+}
+
+// assumes the stateLock is held.
+func (n *network) connectOnce(ip utils.IPDesc) {
 	str := ip.String()
-	if _, ok := n.disconnectedIPs[str]; ok {
+	if !n.shouldTrack(str) {
 		return
 	}
-	if _, ok := n.connectedIPs[str]; ok {
-		return
-	}
-	if _, ok := n.myIPs[str]; ok {
-		return
-	}
-	n.disconnectedIPs[str] = struct{}{}
 
-	go n.connectTo(ip)
+	go func() {
+		err := n.attemptConnect(ip)
+		if err == nil {
```

Comment:
nit: for consistency, can we do the typical `if err != nil` pattern` instead of `if err == nil`?